File: py-methods/hvg_performance.py
import os
import time
import pandas as pd
import scanpy as sc
import argparse
from pathlib import Path
import gc


def parse_args():
    parser = argparse.ArgumentParser(
        description="Benchmark for H5AD operations using Scanpy",
        prog="scanpy_hvg_benchmark"
    )
    parser.add_argument(
        "-i", "--input",
        dest="h5ad_path",
        type=Path,
        required=True,
        help="Path to H5AD file"
    )
    parser.add_argument(
        "-o", "--output",
        dest="csv_path",
        type=Path,
        required=True,
        help="Path for output CSV file"
    )
    parser.add_argument(
        "-n", "--iterations",
        dest="iterations",
        type=int,
        default=5,
        help="Number of benchmark iterations to run"
    )
    return parser.parse_args()


def log_benchmark_result(file_path, step, duration_ms, success, dimensions=None):
    """Log benchmark results to CSV file."""
    
    file_exists = os.path.isfile(file_path) and os.path.getsize(file_path) > 0
    
    dims_str = f"{dimensions[0]},{dimensions[1]}" if dimensions else ","
    
    df = pd.DataFrame({
        "test_case": ["hvg"],
        "test_step": [step],
        "time_ms": [duration_ms],
        "success": [success],
        "rows": [dimensions[0] if dimensions else None],
        "cols": [dimensions[1] if dimensions else None]
    })
    
    df.to_csv(file_path, mode='a', header=not file_exists, index=False)
    

def run_hvg_benchmark(h5ad_path, csv_path, iterations):
    """Run the HVG benchmark with specified parameters."""
    
    print("Starting HVG benchmark")
    
    print(f"Loading dataset from {h5ad_path}")
    load_start = time.time()
    
    try:
        adata = sc.read_h5ad(h5ad_path)
        load_time = int((time.time() - load_start) * 1000) 
        sc.pp.normalize_total(adata, target_sum=1e4)
        sc.pp.log1p(adata)
        load_success = True
        dimensions = (adata.n_obs, adata.n_vars)
        adata_var = adata.var.copy()
        print(f"Dataset loaded: {adata.n_obs} cells x {adata.n_vars} genes")
        
        log_benchmark_result(csv_path, "load", load_time, load_success, dimensions)
        
        print(f"Running HVG computation for {iterations} iterations")
        
        for i in range(1, iterations + 1):
            print(f"  Iteration {i}/{iterations}")
            
            
            hvg_start = time.time()
            try:
                sc.pp.highly_variable_genes(
                    adata, 
                    flavor='seurat',
                    n_bins=20,
                    min_mean=0.0125,
                    max_mean=3,
                    min_disp=0.5,
                    span=0.3
                )
                hvg_success = True
            except Exception as e:
                print(f"HVG computation failed: {e}")
                hvg_success = False
            
            hvg_time = int((time.time() - hvg_start) * 1000) 
            log_benchmark_result(csv_path, "compute_hvg", hvg_time, hvg_success, dimensions)
            adata.var = adata_var.copy()
        print("HVG benchmark completed successfully")
        
    except Exception as e:
        load_time = int((time.time() - load_start) * 1000)  
        load_success = False
        print(f"Failed to load dataset: {e}")
        log_benchmark_result(csv_path, "load", load_time, load_success)
    

def main():
    args = parse_args()
    
    print("Starting HVG benchmark with:")
    print(f"  Input file: {args.h5ad_path}")
    print(f"  Output file: {args.csv_path}")
    print(f"  Iterations: {args.iterations}")
    
    run_hvg_benchmark(args.h5ad_path, args.csv_path, args.iterations)
    
    print("Benchmark completed")


if __name__ == "__main__":
    main()

File: rs-methods/src/hvg_performance.rs
use anndata_memory::DeepClone;
use anyhow::Result;
use clap::Parser;
use rayon::ThreadPoolBuilder;
use single_rust::io;
use single_rust::memory::processing::compute_highly_variable_genes;
use single_rust::shared::Precision;
use single_utilities::types::Direction;
use std::fs::{File, OpenOptions};
use std::io::Write;
use std::path::PathBuf;
use std::time::Instant;

#[derive(Parser)]
#[command(
    author = "Ian Ferenc Diks",
    version,
    about = "Benchmark for H5AD operations"
)]
struct Args {
    /// Path to H5AD file
    #[arg(short = 'i', long = "input")]
    h5ad_path: PathBuf,
    /// Path for output CSV file
    #[arg(short = 'o', long = "output")]
    csv_path: PathBuf,
    /// Number of benchmark iterations to run
    #[arg(short = 'n', long = "iterations", default_value = "5")]
    iterations: usize,
}

fn log_benchmark_result(
    file: &mut File,
    step: &str,
    duration_ms: u128,
    success: bool,
    dimensions: Option<(usize, usize)>,
) -> Result<()> {
    let dims_str = if let Some((obs, vars)) = dimensions {
        format!("{},{}", obs, vars)
    } else {
        String::from(",")
    };

    writeln!(
        file,
        "hvg,{},{},{},{}",
        step, duration_ms, success, dims_str
    )?;
    Ok(())
}

fn run_hvg_benchmark(h5ad_path: &PathBuf, csv_path: &PathBuf, iterations: usize) -> Result<()> {
    println!("Starting HVG benchmark");
    let thread_pool = ThreadPoolBuilder::new().num_threads(128).build()?;

    // Create or open the output CSV file
    let mut file = OpenOptions::new()
        .create(true)
        .append(true)
        .open(csv_path)?;

    // Add header if file is new
    if file.metadata()?.len() == 0 {
        writeln!(file, "test_case,test_step,time_ms,success,rows,cols")?;
    }

    println!("Loading dataset from {}", h5ad_path.display());
    let load_start = Instant::now();
    let adata_result = io::read_h5ad_fast_memory(h5ad_path);
    let load_time = load_start.elapsed().as_millis();
    let load_success = adata_result.is_ok();

    if let Ok(adata) = adata_result {
        log_benchmark_result(
            &mut file,
            "load",
            load_time,
            load_success,
            Some((adata.n_obs(), adata.n_vars())),
        )?;

        let x = adata.x();
        single_rust::memory::processing::normalize_expression(
            &x,
            1e4 as u32,
            &Direction::ROW,
            Some(Precision::Single),
        )?;

        println!(
            "Dataset loaded: {} cells x {} genes",
            adata.n_obs(),
            adata.n_vars()
        );
        println!("Running HVG computation for {} iterations", iterations);

        let var_df = adata.var().deep_clone();

        for i in 1..=iterations {
            println!("  Iteration {}/{}", i, iterations);

            let hvg_start = Instant::now();
            let hvg_result = thread_pool.install(|| compute_highly_variable_genes(&adata, None));
            let hvg_time = hvg_start.elapsed().as_millis();
            let hvg_success = hvg_result.is_ok();

            log_benchmark_result(
                &mut file,
                "compute_hvg",
                hvg_time,
                hvg_success,
                Some((adata.n_obs(), adata.n_vars())),
            )?;
            adata.var().set_data(var_df.get_data())?;
        }

        println!("HVG benchmark completed successfully");
    } else if let Err(e) = adata_result {
        // Log the failed loading step
        log_benchmark_result(&mut file, "load", load_time, load_success, None)?;
        println!("Failed to load dataset: {}", e);
    }

    Ok(())
}

fn main() -> Result<()> {
    let args = Args::parse();

    println!("Starting HVG benchmark with:");
    println!("  Input file: {}", args.h5ad_path.display());
    println!("  Output file: {}", args.csv_path.display());
    println!("  Iterations: {}", args.iterations);

    run_hvg_benchmark(&args.h5ad_path, &args.csv_path, args.iterations)?;

    println!("Benchmark completed");
    Ok(())
}
